zerver/actions/users.py

```diff
@@ -312,6 +317,96 @@ def do_deactivate_user(
+    never_subscribed_private_stream_ids = set()
+
+    if old_role_value == UserProfile.ROLE_GUEST or user_profile.is_guest:
+        # If either the new role or old role was guest, the user will gain or lose
+        # access to never subscribed public streams.
+        all_public_streams = {stream.id for stream in all_streams if stream.is_public()}
+        web_public_stream_ids = {
+            stream.id
+            for stream in all_streams
+            if (stream.is_public() and stream.is_web_public is True)
+        }
+        never_subscribed_public_stream_ids = (
+            all_public_streams - subscription_stream_ids - web_public_stream_ids
+        )
+
+    if user_profile.is_realm_admin or old_role_value in [
+        UserProfile.ROLE_REALM_OWNER,
+        UserProfile.ROLE_REALM_ADMINISTRATOR,
+    ]:
+        # If either the new role or old role was admin or owner, then the user will gain
+        # or lose access to the never subscribed private streams as well.
+        all_private_streams = {stream.id for stream in all_streams if not stream.is_public()}
+        never_subscribed_private_stream_ids = all_private_streams - subscription_stream_ids
+
+    stream_ids_with_changed_permissions = (
+        never_subscribed_public_stream_ids | never_subscribed_private_stream_ids
+    )
+
+    if not stream_ids_with_changed_permissions:
+        # There are no streams whose access permissions have changed for the user due to role update.
+        return
+
+    streams_with_changed_permissions = [
+        all_streams_map[stream_id] for stream_id in stream_ids_with_changed_permissions
+    ]
+    if user_profile.role < old_role_value:
+        # This means that the user is promoted to have higher level of permissions
+        # and we may need to send stream creation events.
+        event_op = "create"
+    else:
+        # This means that the user is demoted to have lower level of permissions
+        # and we may need to send stream deletion events.
+        event_op = "delete"
+
+    event = dict(
+        type="stream",
+        op=event_op,
+        streams=[stream.to_dict() for stream in streams_with_changed_permissions],
+    )
+    send_event_on_commit(user_profile.realm, event, [user_profile.id])
+
+    if event_op == "create":
+        # Need to send peer_add events as well with stream creation events, so
+        # that the user receives subscribers information.
+        subscriber_peer_info = bulk_get_subscriber_peer_info(
+            user_profile.realm, streams_with_changed_permissions
+        )
+        for stream_id, stream_subscriber_set in subscriber_peer_info.subscribed_ids.items():
+            peer_add_event = dict(
+                type="subscription",
+                op="peer_add",
+                stream_ids=[stream_id],
+                user_ids=sorted(stream_subscriber_set),
+            )
+            send_event_on_commit(user_profile.realm, peer_add_event, [user_profile.id])
```

Comment:
We can probably have two separate functions for stream creation and deletion events and let the caller decide what to call based on `user_profile.role < old_role_value` check.